### PR TITLE
Integrating Modular SLAM Pipeline into OpenCV Contrib

### PR DESCRIPTION
**Pull request by QueenofUSSR**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

## Summary

This PR adds the SLAM module (slam) providing a compact visual odometry and small-scale SLAM toolkit. The module includes feature extraction, matching, two-view initialization, pose estimation, local mapping (MapManager / MapPoint), a simple optimizer (local BA with SFM fallback and optional g2o support), visualizer/top-down trajectory, localizer (PnP-based relocalization), and utilities for loading image sequences and intrinsics.

## Motivation

Offer a lightweight, self-contained SLAM implementation that demonstrates how to use OpenCV building blocks for VO/SLAM research and examples. Useful as an educational reference and quick prototyping base for researchers who want to extend or compare with other methods (ORB-SLAM family, etc.). It also provides a straightforward path to enable g2o when available for robust BA.

## What changed / what’s included

`slam.hpp` — module aggregator
`vo.hpp` — VisualOdometry wrapper
`data_loader.hpp` — sequence loader with YAML intrinsics parsing
`feature.hpp` — FeatureExtractor (ORB + ANMS + flow-aware selection)
`matcher.hpp` — Matcher (ratio test, bucketing, mutual check)
`initializer.hpp` — Two-view initializer using H/F decomposition + triagulation
`pose.hpp` — PoseEstimator (essential matrix + recoverPose)
`map.hpp` & `keyframe.hpp` — MapManager, MapPoint, KeyFrame
`localizer.hpp` — PnP-based relocalizer
`optimizer.hpp` — Optimizer (local BA using either g2o or OpenCV-based SFM fallback)
`visualizer.hpp` — simple frame/top-down visualization and trajectory saving
Implementation (under src/):
`vo.cpp` — main VisualOdometry run loop, backend thread for local BA, CSV diagnostics, initialization + tracking + triangulation pipeline
`data_loader.cpp` — filesystem/glob image enumerator, optional sensor.yaml intrinsics parsing (small comment fix in this PR)
`feature.cpp`, `matcher.cpp`, `initializer.cpp`, `pose.cpp`, `optimizer.cpp`, `visualizer.cpp`, `localizer.cpp` — corresponding implementations described above

### CODE REVIEW

**Comment by abhishek-gola** [comment]
```
FAILED: modules/slam/CMakeFiles/opencv_slam.dir/src/data_loader.cpp.o 
/usr/bin/ccache /usr/bin/c++  -DCVAPI_EXPORTS -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/ci/opencv/hal/ipp/include -I3rdparty/ippicv/ippicv_lnx/icv/include -I3rdparty/ippicv/ippicv_lnx/iw/include -I/home/ci/opencv/3rdparty/dlpack/include -I/home/ci/opencv_contrib/modules/slam/include -Imodules/slam -I/home/ci/opencv_contrib/modules/cudev/include -I/home/ci/opencv/modules/core/include -I/home/ci/opencv_contrib/modules/cudaarithm/include -I/home/ci/opencv/modules/flann/include -I/home/ci/opencv/modules/imgproc/include -I/home/ci/opencv/modules/ml/include -I/home/ci/opencv/modules/dnn/include -I/home/ci/opencv/modules/features2d/include -I/home/ci/opencv/modules/imgcodecs/include -I/home/ci/opencv/modules/videoio/include -I/home/ci/opencv/modules/calib3d/include -I/home/ci/opencv/modules/highgui/include -I/home/ci/opencv_contrib/modules/shape/include -I/home/ci/opencv/modules/video/include -I/home/ci/opencv_contrib/modules/xfeatures2d/include -I/home/ci/opencv_contrib/modules/sfm/include -isystem . -isystem /usr/include/eigen3 -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse3 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG  -DNDEBUG -fPIC   -std=c++17 -MD -MT modules/slam/CMakeFiles/opencv_slam.dir/src/data_loader.cpp.o -MF modules/slam/CMakeFiles/opencv_slam.dir/src/data_loader.cpp.o.d -o modules/slam/CMakeFiles/opencv_slam.dir/src/data_loader.cpp.o -c /home/ci/opencv_contrib/modules/slam/src/data_loader.cpp
/home/ci/opencv_contrib/modules/slam/src/data_loader.cpp:7:6: error: invalid preprocessing directive #namespace
    7 | #    namespace fs = std::filesystem;
      |      ^~~~~~~~~
/home/ci/opencv_contrib/modules/slam/src/data_loader.cpp: In constructor 'cv::vo::DataLoader::DataLoader(const string&)':
/home/ci/opencv_contrib/modules/slam/src/data_loader.cpp:33:9: error: 'fs' has not been declared
   33 |         fs::path p(imageDir);
      |         ^~
/home/ci/opencv_contrib/modules/slam/src/data_loader.cpp:34:13: error: 'fs' has not been declared
   34 |         if(!fs::exists(p) || !fs::is_directory(p)){
      |             ^~
/home/ci/opencv_contrib/modules/slam/src/data_loader.cpp:34:24: error: 'p' was not declared in this scope
   34 |         if(!fs::exists(p) || !fs::is_directory(p)){
      |                        ^
/home/ci/opencv_contrib/modules/slam/src/data_loader.cpp:34:31: error: 'fs' has not been declared
   34 |         if(!fs::exists(p) || !fs::is_directory(p)){
      |                               ^~
/home/ci/opencv_contrib/modules/slam/src/data_loader.cpp:67:5: error: 'fs' has not been declared
   67 |     fs::path p(imageDir);
      |     ^~
/home/ci/opencv_contrib/modules/slam/src/data_loader.cpp:68:26: error: 'p' was not declared in this scope
   68 |     std::string yaml1 = (p / "sensor.yaml").string();
      |                          ^
[2987/3796] Building CXX object modules/sfm/src/libmv/libmv/correspondence/CMakeFiles/opencv.sfm.correspondence.dir/nRobustViewMatching.cc.o
[2988/3796] Building CXX object modules/slam/CMakeFiles/opencv_slam.dir/src/matcher.cpp.o
/home/ci/opencv_contrib/modules/slam/src/matcher.cpp: In member function 'void cv::vo::Matcher::match(const cv::Mat&, const cv::Mat&, const std::vector<cv::KeyPoint>&, const std::vector<cv::KeyPoint>&, std::vector<cv::DMatch>&, int, int, int, int, int, int, bool, bool)':
/home/ci/opencv_contrib/modules/slam/src/matcher.cpp:26:50: warning: unused parameter 'kps1' [-Wunused-parameter]
   26 |                     const std::vector<KeyPoint> &kps1, const std::vector<KeyPoint> &kps2,
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
[2989/3796] Building CXX object modules/slam/CMakeFiles/opencv_slam.dir/src/feature.cpp.o
/home/ci/opencv_contrib/modules/slam/src/feature.cpp: In lambda function:
/home/ci/opencv_contrib/modules/slam/src/feature.cpp:142:80: warning: declaration of 'b' shadows a previous local [-Wshadow]
  142 |         std::sort(b.begin(), b.end(), [](const CandScore &a, const CandScore &b){ return a.score > b.score; });
      |                                                                                ^
/home/ci/opencv_contrib/modules/slam/src/feature.cpp:140:15: note: shadowed declaration is here
  140 |     for(auto &b: buckets){
      |               ^
[2990/3796] Building CXX object modules/slam/CMakeFiles/opencv_slam.dir/src/localizer.cpp.o
[2991/3796] Building CXX object modules/slam/CMakeFiles/opencv_slam.dir/src/map.cpp.o
[2992/3796] Building CXX object modules/slam/CMakeFiles/opencv_slam.dir/src/initializer.cpp.o
/home/ci/opencv_contrib/modules/slam/src/initializer.cpp: In member function 'void cv::vo::Initializer::decomposeH(const cv::Mat&, std::vector<cv::Mat>&, std::vector<cv::Mat>&, std::vector<cv::Mat>&)':
/home/ci/opencv_contrib/modules/slam/src/initializer.cpp:318:9: warning: unused variable 'solutions' [-Wunused-variable]
  318 |     int solutions = decomposeHomographyMat(H_normalized, Mat::eye(3,3,CV_64F),
      |         ^~~~~~~~~
```

@QueenofUSSR build is failing.

**Comment by abhishek-gola** [comment]
@QueenofUSSR, can you add some tests and a sample?

**Comment by QueenofUSSR** [comment]
> @QueenofUSSR, can you add some tests and a sample?

Okay, we will upload some tests and a visualized sample soon!

File: modules/slam/src/data_loader.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                    if(stat(cur.c_str(), &st) == 0 && S_ISDIR(st.st_mode)){
+                        // ok
+                    } else return false;
+                }
+            }
+        }
+        if(next == std::string::npos) break;
+        pos = next + 1;
+    }
+    return true;
```
Comment (abhishek-gola) [inline modules/slam/src/data_loader.cpp]: Use opencv's filesystem `<opencv2/core/utils/filesystem.hpp>` it has inbuilt functions like **createDirectories** and **isDirectory**

File: modules/slam/src/initializer.cpp [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+    points3D.resize(pts1.size());
+    isTriangulated.resize(pts1.size(), false);
+
+    Mat P1 = Mat::eye(3, 4, CV_64F);
+    Mat P2(3, 4, CV_64F);
+    for(int i = 0; i < 3; ++i) {
+        for(int j = 0; j < 3; ++j) P2.at<double>(i,j) = R.at<double>(i,j);
+        P2.at<double>(i, 3) = t.at<double>(i, 0);
+    }
+
```
Comment (abhishek-gola) [inline modules/slam/src/initializer.cpp]:
```
cv::Mat P2;
cv::hconcat(R, t, P2);
```
It can also be done this way.

File: modules/slam/src/vo.cpp [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+                    if(t < 0 || t >= (int)rev.size() || rev[t].empty()) continue;
+                    DMatch rbest = rev[t][0];
+                    if((forward && rbest.trainIdx == (int)qi) || (!forward && rbest.trainIdx == best.queryIdx)){
+                        filtered.push_back(best);
+                    }
+                }
+                return filtered;
+            };
+            std::vector<DMatch> goodMatches = ratioKeep(knn12, true);
+
```
Comment (abhishek-gola) [inline modules/slam/src/vo.cpp]: why do we need this lambda function? you can use the code directly.

**Comment by abhishek-gola** [comment]
@QueenofUSSR, Please let me know once you are done adding tests and sample.

**Comment by QueenofUSSR** [comment]
> @QueenofUSSR, Please let me know once you are done adding tests and sample.

Thank you for your notification! Now there is a simple visualized VO sample. It would be built at [build directory]/bin/example_slam_run_vo_sample if `-DBUILD_EXAMPLES=ON` is set. It has been tested with EuRoC MH01 and several video sequences we made. My partner and I have just finished our final exam and we will conduct further tests these days.

**Comment by QueenofUSSR** [comment]
@abhishek-gola We have tested our VO module on EuRoC MH01 cam0, Root Mean Square of Absolute Trajectory Error is 2.46924m, Root Mean Square of Relative Pose Error is 0.0367811m. The test code is in `module/slam/test/`. We are trying to find improvements.
